---
title: "Check Library Sizes"
author: "Steve Pederson"
date: "11 December 2018"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,
											message = FALSE, warning = FALSE)
```

```{r}
library(ngsReports)
library(edgeR)
library(magrittr)
library(scales)
library(tidyverse)
library(ggrepel)
library(pander)
theme_set(theme_bw())
```

## Load all FastQC summaries

```{r}
rawFqc <- list.files("../0_rawData/FastQC/", pattern = "zip$", full.names = TRUE) %>%
	getFastqcData()
trimFqc <- list.files("../1_trimmedData/FastQC/", pattern = "zip$", full.names = TRUE) %>%
	getFastqcData()
alnFqc <- list.files("../2_alignedData/FastQC/", pattern = "zip$", full.names = TRUE) %>%
	getFastqcData()
```

### Read Totals

```{r}
rt <- list(
	readTotals(rawFqc) %>% mutate(Stage = "Raw"),
	readTotals(trimFqc) %>% mutate(Stage = "Trimmed"),
	readTotals(alnFqc) %>% mutate(Stage = "Aligned")
) %>%
	bind_rows() %>%
	mutate(Sample = str_remove_all(Filename, "(_R1.fq.gz|Aligned.sortedByCoord.out.bam)"),
				 Sample = factor(Sample, levels = str_sort(unique(Sample), numeric = TRUE)),
				 Stage = factor(Stage, levels = c("Raw", "Trimmed", "Aligned")))
```

```{r, echo=FALSE, fig.cap="*Summary of library sizes after each step. The slight increase after alignments indicates a low level of multiple alignments*"}
plotly::ggplotly(
rt %>%
	ggplot(aes(Sample, Total_Sequences, fill = Stage)) +
	geom_bar(stat = "identity", position = "dodge") +
	scale_y_continuous(labels = scales::comma, 
										 expand = expand_scale(mult = c(0, 0.05))) +
	theme_bw() +
	theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1))
)
```

- Multiple alignments don't appear to be much of an issue
- Trimming made little change to the library sizes
- All samples were run in technical duplicates which can be combined for DE analysis

### GC Content

No unusual behaviour was found in GC content.

```{r, echo=FALSE, fig.cap="*GC content for raw data*"}
plotGcContent(rawFqc, plotType = "line", species = "Drerio", theoreticalType = "transcriptome")
```

```{r, echo=FALSE, fig.cap="*GC content for aligned data*"}
plotGcContent(alnFqc, plotType = "line", species = "Drerio", theoreticalType = "transcriptome", pattern = "Aligned.+")
```

### Duplication Levels

No issues were evident with regard to duplication levels

```{r, echo=FALSE, fig.cap = "*Duplication levels for raw data*"}
plotDuplicationLevels(rawFqc, pattern = "_R1.+")
```

```{r, echo=FALSE, fig.cap = "*Duplication levels for aligned data*"}
plotDuplicationLevels(alnFqc, pattern = "Aligned.+")
```

## Alignment Rate

Alignment rates were also very good

```{r}
alnLogs <- list.files("../2_alignedData/logs/", full.names = TRUE, pattern = "final.out") %>%
	importStarLogs()
```

```{r, echo=FALSE, fig.cap="*Alignment summaries for all libraries (before merging replicates)*"}
alnLogs %>%
	rename(Library = Filename) %>%
  dplyr::select(Library, 
                Number_Of_Input_Reads,
                Uniquely_Mapped_Reads_Number, 
                Number_Of_Reads_Mapped_To_Multiple_Loci,
                Number_Of_Reads_Mapped_To_Too_Many_Loci) %>%
  mutate(Unmapped = Number_Of_Input_Reads - Uniquely_Mapped_Reads_Number - Number_Of_Reads_Mapped_To_Multiple_Loci - Number_Of_Reads_Mapped_To_Too_Many_Loci) %>%
  gather("Type", "Total", -Library, -Number_Of_Input_Reads) %>%
  mutate(Library = str_remove_all(Library, "Log.final.out"),
  			 Library = str_replace(Library, "__-", "_WT"),
  			 Genotype = str_extract(Library, "(FAD|FS|WT)"),
         Type = str_remove_all(Type, "(Number_Of_Reads_|_Reads_Number)"),
         Type = factor(Type, levels = c("Unmapped", "Mapped_To_Too_Many_Loci", "Mapped_To_Multiple_Loci", "Uniquely_Mapped")),
         Percent = percent(Total / Number_Of_Input_Reads),
         Number_Of_Reads = comma(Total)) %>%
  ggplot(aes(Library, Total, fill = Type, percent = Percent)) +
  geom_bar(stat = "identity") +
  facet_grid(~Genotype, scales = "free_x", space = "free_x") +
  scale_y_continuous(labels = scales::comma, 
                     expand = expand_scale(c(0, 0.05))) +
  scale_fill_manual(values = c("grey50", "red", "yellow", rgb(0, 0.8, 0.15))) +
  labs(x = "Library",
       y = "Total Reads") +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1))
```


## Gene-level Counts

### MDS Plot

Gene-level counts were loaded without filtering and an MDS plot generated in order to check whether replicate libraries grouped correctly.

```{r}
counts <- read_tsv("../2_alignedData/featureCounts/genes.out") %>%
	set_names(basename(colnames(.))) %>%
	set_names(str_remove_all(colnames(.), "Aligned.sortedByCoord.out.bam"))
```

```{r}
dge <- counts %>% 
	as.data.frame() %>%
	column_to_rownames("Geneid") %>%
	DGEList() %>%
	calcNormFactors()
dge$samples %<>%
	rownames_to_column("library") %>%
	mutate(Genotype = str_extract_all(library, "(FAD|FS|__)"),
				 Genotype = str_replace_all(Genotype, "__", "WT"),
				 Genotype = factor(Genotype, levels = c("WT", "FS", "FAD")),
				 Sample = str_replace(library, "__-", "_WT"),
				 Sample = ifelse(str_count(Sample, "_") == 3, Sample, paste0(Sample, "_1")),
				 Replicate = str_extract(Sample, "[12]$"),
				 Sample = str_remove(Sample, "_[12]$"),
				 group = as.integer(Genotype)) %>%
	column_to_rownames("library")
```

```{r}
mds <- plotMDS(dge, plot = FALSE) %>%
	extract2("cmdscale.out") %>%
	set_colnames(paste0("Dim", 1:2)) %>%
	as.data.frame() %>%
	rownames_to_column("Library") %>%
	as_tibble() 
```


```{r, echo=FALSE, fig.cap = "*Plot of replicate libraries showing good concordance between replicates. These will be combined and low-expressed genes removed for the DE analysis*"}
mds %>%
	cbind(dge$samples[.$Library,]) %>%
	ggplot(aes(Dim1, Dim2, colour = Genotype, label = Sample)) +
	geom_point() +
	geom_text_repel()
```

### Library sizes

It was noted that library sizes were about 10-20 fold smaller than expected after summarising to gene level counts and these ranged between `r pander(comma(range(dge$samples$lib.size)))`.
Alignments were then inspected manually in order to ascertain the cause of this low counting rate.