---
title: "Check Library Sizes"
author: "Steve Pederson"
date: "11 December 2018"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}
library(ngsReports)
library(edgeR)
library(tidyverse)
```

## Load all FastQC summaries

```{r}
rawFqc <- list.files("../0_rawData/FastQC/", pattern = "zip$", full.names = TRUE) %>%
	getFastqcData()
trimFqc <- list.files("../1_trimmedData/FastQC/", pattern = "zip$", full.names = TRUE) %>%
	getFastqcData()
alnFqc <- list.files("../2_alignedData/FastQC/", pattern = "zip$", full.names = TRUE) %>%
	getFastqcData()
```

## Read Totals

```{r}
rt <- list(
	readTotals(rawFqc) %>% mutate(Stage = "Raw"),
	readTotals(trimFqc) %>% mutate(Stage = "Trimmed"),
	readTotals(alnFqc) %>% mutate(Stage = "Aligned")
) %>%
	bind_rows() %>%
	mutate(Sample = str_remove_all(Filename, "(_R1.fq.gz|Aligned.sortedByCoord.out.bam)"),
				 Sample = factor(Sample, levels = str_sort(unique(Sample), numeric = TRUE)),
				 Stage = factor(Stage, levels = c("Raw", "Trimmed", "Aligned")))
```

```{r}
rt %>%
	ggplot(aes(Sample, Total_Sequences, fill = Stage)) +
	geom_bar(stat = "identity", position = "dodge") +
	scale_y_continuous(labels = scales::comma) +
	theme_bw() +
	theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1))
```

- Multiple aligments don't appear to be much of an issue
- Trimming made little change to the library sizes
- All samples were run in tehcnical duplicates which can be combined for DE analysis

# Gene-level Counts

```{r}
counts <- read_tsv("../2_alignedData/featureCounts/genes.out") %>%
	set_colnames(basename(colnames(.))) %>%
	set_colnames(str_remove_all(colnames(.), "Aligned.sortedByCoord.out.bam"))
```

```{r}
dge <- counts %>% 
	as.data.frame() %>%
	column_to_rownames("Geneid") %>%
	DGEList()
```

```{r}
mds <- plotMDS(dge, plot = FALSE) %>%
	extract2("cmdscale.out") %>%
	set_colnames(paste0("Dim", 1:2)) %>%
	as.data.frame() %>%
	rownames_to_column("Library") %>%
	as_data_frame() 
```

Check replicate samples
