---
title: "Differential Gene Expression Analysis"
author: "Steve Pederson"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output: 
  html_document: 
    fig_caption: yes
    fig_height: 6
    fig_width: 8
    toc: yes
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,
											message = FALSE, warning = FALSE)
```

## Setup

```{r loadPackages}
library(limma)
library(edgeR)
library(AnnotationHub)
library(magrittr)
library(scales)
library(tidyverse)
library(pander)
theme_set(theme_bw())
```

The alignments in this analysis were generated by aligning each library (including technical replicates) to the Zebrafish genome GRCz11 (Ensembl Release 94) using (STAR v2.5.3a).
Reads were assigned to each genes based on 100% exonic overlap and unique genomic alignments, based on gene descriptions in `Danio_rerio.GRCz11.94.chr.gtf`, also obtained from Ensembl Release 94.
This set of gene descriptions were then loaded into `R` as an `EnsDb` object using the `AnnotationHub()` infrastructure.

```{r genes}
ah <- AnnotationHub() %>%
	subset(species == "Danio rerio") %>%
	subset(rdataclass == "EnsDb")
ensDb <- ah[["AH64906"]]
genes <- genes(ensDb)
```


```{r counts}
counts <- read_tsv("../2_alignedData/featureCounts/genes.out") %>%
	set_names(basename(colnames(.))) %>%
	set_names(str_remove_all(colnames(.), "Aligned.sortedByCoord.out.bam")) 
```

```{r, echo=FALSE, fig.cap="*Total counts from each library after assigning to genes*"}
counts %>% 
	as.data.frame() %>%
	column_to_rownames("Geneid") %>%
	colSums() %>%
	as.data.frame() %>%
	set_names("Counts") %>%
	rownames_to_column("Library") %>%
	as_tibble() %>%
	mutate(Library = str_replace_all(Library, "__-", "_WT"),
				 Sample = case_when(
				 	str_count(Library, "_") == 2 ~ paste0(Library, "_1"),
				 	str_count(Library, "_") == 3 ~ Library
				 ),
				 Replicate = str_extract(Sample, "[12]$"),
				 Sample = str_remove_all(Sample, "_[12]$"),
				 Genotype = str_extract(Sample, "(FAD|FS|WT)")) %>%
	ggplot(aes(Sample, Counts, fill = Replicate)) +
	geom_bar(stat = "identity") +
	scale_y_continuous(labels = comma, expand = expand_scale(c(0, 0.05))) +
	scale_fill_manual(values = RColorBrewer::brewer.pal(3, "Set1")[1:2]) +
	facet_wrap(~Genotype, scales = "free_x") +
	theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5))
```

```{r dge}
minCpm <- 1
minSamples <- 5
dge <- counts %>%
	gather(key = "Library", value = "Counts", -Geneid) %>%
	mutate(Library = str_replace_all(Library, "__-", "_WT"),
				 Sample = case_when(
				 	str_count(Library, "_") == 2 ~ paste0(Library, "_1"),
				 	str_count(Library, "_") == 3 ~ Library
				 ),
				 Sample = str_remove_all(Sample, "_[12]$")) %>%
	dplyr::filter(Counts > 0) %>%
	group_by(Geneid, Sample) %>%
	summarise(Counts = sum(Counts)) %>%
	spread(key = Sample, value = Counts, fill = 0) %>%
	as.data.frame() %>%
	column_to_rownames("Geneid") %>%
	magrittr::extract(rowSums(cpm(.) > minCpm) >= minSamples,) %>%
	DGEList(genes = genes[rownames(.)]) %>%
	calcNormFactors()
```

Extracted RNA from each individual zebrafish was split and run across two NextSeq lanes, giving technical replicates for the sequencing step only.
These were combined to provide a single library for each sample, as technical replication at this experimental step is not of particular interest.
Genes were retained for analysis if a CPM > `r minCpm` was observed for $\geq$ `r minSamples` samples.
This equated to about `r floor(mean(minCpm*dge$samples$lib.size/1e6))` reads for a gene in at least `r minSamples` samples for inclusion in downstream analysis, giving a total of `r comma(nrow(dge))` of the original `r comma(nrow(counts))` genes for DGE analysis.

```{r}
dge$samples %<>%
	mutate(Sample = rownames(.),
				 Genotype = str_extract(Sample, "(WT|FAD|FS)"),
				 Genotype = factor(Genotype, levels = c("WT", "FS", "FAD")),
				 group = as.integer(Genotype)) %>%
	set_rownames(.$Sample)
```

```{r voom}
v <- voomWithQualityWeights(dge, design = matrix(rep(1, ncol(dge)), ncol = 1))
```

Counts were also processed using the `voom` transformation using quality weights to allow for analysis using normal-based algorithms.
Samples weights ranged between `r pander(round(range(v$targets$sample.weights), 4))`, with the most down-weigted sample being a WT sample.

```{r, echo=FALSE}
v$targets %>% 
	ggplot(aes(Sample, sample.weights, fill = Genotype)) + 
	geom_bar(stat = "identity") + 
	facet_wrap(~Genotype, scales= "free_x") + 
	geom_hline(yintercept = 1, linetype = 2) +
	scale_y_continuous(expand = expand_scale(c(0, 0.05))) +
	labs(y = "Sample Weight") +
	theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5))
```


## Analysis

### Data Inspection

The first step for the analysis was to perform an MDS analysis.
However, minimal separation was observed between sample groups, with the exception of the most strongly down-weighted sample, which separated very clearly from the remainder of the points.
This sample may be a candidate for exclusion if no differentially expressed genes are able to be detected.
A simple PCA also revealed that the first few principal components only capture a lesser amount of the total variability than may be expected.

```{r mds}
mds <- plotMDS(v, plot = FALSE) %>%
	extract2("cmdscale.out") %>%
	set_colnames(paste0("Dim", 1:2))
```

```{r, echo=FALSE, fig.cap = "*MDS plot showing no clear groups within the data. Point sizes indicate sample weights as calculated by `voomWithQualityWeights()`.*"}
plotly::ggplotly(mds %>%
	cbind(dge$samples[rownames(.),]) %>%
	cbind(w = v$targets[rownames(.), "sample.weights"]) %>%
	ggplot(aes(Dim1, Dim2, colour = Genotype, size = w, label = Sample)) +
	geom_point() +
	guides(size = FALSE)
)
```

```{r, echo=FALSE}
dge %>%
	cpm(log = TRUE) %>%
	t() %>%
	prcomp() %>%
	summary() %>%
	extract2("importance") %>%
	magrittr::extract(,paste0("PC", 1:5)) %>%
	pander(caption = paste("*First five principal components, showing that the first two only account for", percent(.[3,2]), "of the total variance.*"))
```

### Genotype checks

In order to confirm the genotypes of all libraries, three seed sequences were searched in the trimmed reads (i.e. before alignment) which represented the WT, FAD or FS genotypes.
The number of matches to each of the seeds was counted for each sample as a simple strategy for confirming genotypes.

```{r, echo=FALSE}
readLines("../bash/checkPsen2.sh") %>%
	grep(pattern = "^(WT|FAD|FS)", x = ., value = TRUE) %>% 
	as.data.frame() %>% 
	set_colnames("X") %>% 
	mutate(X = gsub("=", "_", X)) %>% 
	separate(X, into = c("Genotype", "Seed")) %>% 
	pander(caption = "Genotypes & Seed sequences use for genotype checking")
```

```{r, echo=FALSE}
read_tsv("../1_trimmedData/psen2checks/psen.txt") %>%
	mutate(Filename = str_remove(Filename, "_R1.fq.gz"),
				 Filename = str_replace(Filename, "__-", "_WT"),
				 Sample = str_extract(Filename, "[0-9]+_[A-Z]+_[0-9]+")) %>%
	group_by(Sample) %>% 
	summarise_at(c("WT", "FAD", "FS"), sum) %>%
	left_join(dge$samples) %>%
	mutate(Observed = case_when(
		FAD == 0 & FS == 0 ~ "WT",
		FAD > 0 & FS == 0 ~ "FAD",
		FAD == 0 & FS > 0 ~ "FS"
	)) %>%
	dplyr::select(Sample, WT, FAD, FS, Genotype, Observed) %>%
	mutate(Confirmed = Observed == Genotype) %>%
	pander(caption = "Results from genotype checks based on seed sequences")
```

One sample (`8_FS_4`) was labelled as an FS sample but presented with the contradictory FAD genotype and this was investgated manually.
Notably, no WT or FS reads were found in this sample using this simple strategy.

```{r}
dge$samples["8_FS_4","Genotype"] <- "FAD"
v$targets["8_FS_4","Genotype"] <- "FAD"
```


### Comparisons

```{r}
mm <- model.matrix(~0 + Genotype, data = v$targets) %>%
	set_colnames(gsub("Genotype", "", colnames(.)))
cm <- makeContrasts(FSvWT = FS - WT,
										FADvWT = FAD - WT,
										FADvFS = FAD - FS,
										levels = colnames(mm))
```

Three comparisons were defined with the first two being the difference between the two mutant families and the wild-type samples.
The third comparison was defined as being between the two mutant groups.

```{r}
vFit <- v %>%
	lmFit(mm) %>%
	contrasts.fit(cm) %>%
	eBayes()
```

```{r, eval=FALSE}
colnames(cm) %>%
	sapply(function(x){
		topTable(vFit, x) %>%
			mutate(Comparison = x)
	}, simplify = FALSE)
```


### FS Vs WT

<!-- Is 4_FAD a WT sample? 1, 2, 3 & 6 all check out, but the C-C / G -G hetrozygous structure is not evident in this sample-->