---
title: "Differential Gene Expression Analysis"
author: "Steve Pederson"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output: 
  html_document: 
    fig_caption: yes
    fig_height: 6
    fig_width: 8
    toc: yes
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,
											message = FALSE, warning = FALSE)
```

## Setup

```{r}
library(limma)
library(edgeR)
library(AnnotationHub)
library(magrittr)
library(scales)
library(tidyverse)
theme_set(theme_bw())
```

The alignments in this analysis were generated by aligning each library (including technical replicates) to the Zebrafish genome GRCz11 (Ensembl Release 94) using (STAR v2.5.3a).
Reads were assigned to each genes based on 100% exonic overlap and unique genomic alignments, based on gene descriptions in `Danio_rerio.GRCz11.94.chr.gtf`, also obtained from Ensembl Release 94.
This set of gene descriptions were then loaded into `R` as an `EnsDb` object using the `AnnotationHub()` infrastructure.

```{r}
ah <- AnnotationHub() %>%
	subset(species == "Danio rerio") %>%
	subset(rdataclass == "EnsDb")
ensDb <- ah[["AH64906"]]
genes <- genes(ensDb)
```


Extracted RNA from each individual zebrafish was split and run across two NextSeq lanes, giving technical replicates for the sequencing step only.
These were combined to provide a single library for each sample, as technical replication at this step is not of particular interest.

```{r}
counts <- read_tsv("../2_alignedData/featureCounts/genes.out") %>%
	set_names(basename(colnames(.))) %>%
	set_names(str_remove_all(colnames(.), "Aligned.sortedByCoord.out.bam")) 
```

```{r, echo=FALSE, fig.cap="*Total counts from each library*"}
counts %>% 
	as.data.frame() %>%
	column_to_rownames("Geneid") %>%
	colSums() %>%
	as.data.frame() %>%
	set_names("Counts") %>%
	rownames_to_column("Library") %>%
	as_tibble() %>%
	mutate(Library = str_replace_all(Library, "__-", "_WT"),
				 Sample = case_when(
				 	str_count(Library, "_") == 2 ~ paste0(Library, "_1"),
				 	str_count(Library, "_") == 3 ~ Library
				 ),
				 Replicate = str_extract(Sample, "[12]$"),
				 Sample = str_remove_all(Sample, "_[12]$"),
				 Genotype = str_extract(Sample, "(FAD|FS|WT)")) %>%
	ggplot(aes(Sample, Counts, fill = Replicate)) +
	geom_bar(stat = "identity") +
	scale_y_continuous(labels = comma, expand = expand_scale(c(0, 0.05))) +
	scale_fill_manual(values = RColorBrewer::brewer.pal(3, "Set1")[1:2]) +
	facet_wrap(~Genotype, scales = "free_x") +
	theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5))
```

