---
title: "Differential Gene Expression Analysis"
author: "Steve Pederson"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output: 
  html_document: 
    fig_caption: yes
    fig_height: 6
    fig_width: 8
    toc: yes
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE,
											message = FALSE, warning = FALSE)
```

## Setup

```{r loadPackages}
library(limma)
library(edgeR)
library(tximport)
library(AnnotationHub)
library(magrittr)
library(scales)
library(tidyverse)
library(ggrepel)
library(fgsea)
library(pander)
library(here)
```

```{r otherOptions}
if (interactive()) setwd(here("R"))
panderOptions("table.split.table", Inf)
theme_set(theme_bw())
```

The alignments in this analysis were generated by aligning each library (including technical replicates) to the Zebrafish transcriptome from Ensembl Release 94 (GRCz11) using kallisto (v0.43.1).
In addition to the standard transcriptome, the two mutant *psen2* transcripts were manually added to the reference.

The corresponding set of gene descriptions were then loaded into `R` as an `EnsDb` object using the `AnnotationHub()` infrastructure.
Likewise, the set of transcript descriptions were loaded, with the manual addition of the two novel *psen2* mutants.

```{r genes}
ah <- AnnotationHub() %>%
	subset(species == "Danio rerio") %>%
	subset(rdataclass == "EnsDb")
ensDb <- ah[["AH64906"]]
genes <- genes(ensDb) 
mcols(genes) <- mcols(genes)[!names(mcols(genes)) %in% c("seq_coord_system", "gene_name")]
```

```{r transcripts}
transcripts <- transcripts(ensDb)
psen2 <- subset(transcripts, gene_id == "ENSDARG00000015540") %>% 
	granges() %>% 
	magrittr::extract(c(1, 1),) 
names(psen2) <- c("psen2T141_L142delinsMISLISV", "psen2N140fs")
mcols(psen2) <- DataFrame(tx_id = names(psen2),
													tx_biotype = c("protein_coding", "nonsense_mediated_decay"),
													gene_id = "ENSDARG00000015540",
													tx_id_version = names(psen2),
													tx_name = names(psen2))
transcripts %<>%
	GRangesList(psen2) %>%
	unlist() %>%
	sort()
transcripts$symbol <- mcols(genes[transcripts$gene_id])$symbol
# tx2gene <- mcols(transcripts)[c("tx_id_version", "gene_id")] 
tx2gene <- mcols(transcripts)[c("tx_id_version", "symbol")] %>%
	subset(symbol != "")
```


### Gene-level Counts

```{r geneCounts}
geneCounts <- list.files("../3_kallisto/", recursive = TRUE, pattern = "h5", full.names = TRUE) %>%
	set_names(basename(dirname(.))) %>%
	set_names(str_replace(names(.), "_-", "WT")) %>%
	tximport(type = "kallisto", tx2gene = tx2gene)
```

Gene-level counts were imported using `tximport`, mapping transcripts to genes.
Some genes exist in the primary assembly and on alternate assemblies for specific regions, and these were considered as separate transcripts of the same gene for read summarisation purposes. 
Transcript counts were thus mapped to genes using the gene symbol (e.g. *psen2*), instead of the gene id.

```{r dge}
minCpm <- 1
minSamples <- 5
dge <- geneCounts$counts %>%
	magrittr::extract(rowSums(cpm(.) > minCpm) >= minSamples,) %>%
	DGEList() %>%
	calcNormFactors()
dge$samples %<>%
	mutate(Sample = rownames(.),
				 Genotype = str_extract(Sample, "(WT|FAD|FS)"),
				 Genotype = factor(Genotype, levels = c("WT", "FS", "FAD")),
				 group = as.integer(Genotype)) %>%
	set_rownames(.$Sample)
```

Genes were retained for analysis if a CPM > `r minCpm` was observed for $\geq$ `r minSamples` samples.
This equated to about `r floor(mean(minCpm*dge$samples$lib.size/1e6))` reads for a gene in at least `r minSamples` samples for inclusion in downstream analysis, giving a total of `r comma(nrow(dge))` of the original `r comma(nrow(counts))` genes for DGE analysis.

```{r, echo=FALSE, fig.cap="*Total counts from each library after assigning to genes*", fig.height=4}
dge$samples %>% 
	ggplot(aes(Sample, lib.size, fill = Genotype)) +
	geom_bar(stat = "identity") +
	scale_y_continuous(labels = comma, expand = expand_scale(c(0, 0.05))) +
	facet_wrap(~Genotype, scales = "free_x") +
	labs(y = "Library Size") +
	theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5))
```

```{r geneVoom}
geneVoom <- dge %>%
	voomWithQualityWeights(design = matrix(rep(1, ncol(.)), ncol = 1))
```

Counts were also processed using the `voom` transformation using quality weights to allow for analysis using normal-based algorithms.
Sample weights ranged between `r pander(round(range(geneVoom$targets$sample.weights), 4))`, with the most strongly down-weighted being a WT sample.

```{r, echo=FALSE, fig.height=4}
geneVoom$targets %>% 
	ggplot(aes(Sample, sample.weights, fill = Genotype)) + 
	geom_bar(stat = "identity") + 
	facet_wrap(~Genotype, scales= "free_x") + 
	geom_hline(yintercept = 1, linetype = 2) +
	scale_y_continuous(expand = expand_scale(c(0, 0.05))) +
	labs(y = "Sample Weight") +
	theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5))
```

### Transcript-level Counts

Transcript-level counts were imported using `catchKallisto()` from `edgeR` in order to utilise the voom transformation on transcript-level counts.

```{r transCounts, results='hide'}
transCounts <- list.files("../3_kallisto/", full.names = TRUE) %>%
	catchKallisto()
colnames(transCounts$counts) %<>%
	basename() %>%
	str_replace("_-", "WT")
```

```{r transVoom, fig.show='hide'}
transVoom <- transCounts$counts %>%
	magrittr::extract(rowSums(cpm(.) > minCpm) >= minSamples,) %>%
	divide_by(transCounts$annotation[rownames(.),"Overdispersion"]) %>%
	set_rownames(str_remove(rownames(.), "\\.[0-9]+")) %>%
	DGEList(samples = dge$samples[colnames(.),],
					genes = transcripts[rownames(.)]) %>%
	calcNormFactors() %>%
	voomWithQualityWeights(plot = TRUE, design = matrix(1, nrow = ncol(.), ncol = 1))
```

```{r, echo=FALSE, fig.cap = "*Sample weights using transcript-level counts, showing near identical patterns to those observed at the gene-level.*", fig.height=4}
transVoom$targets %>% 
	ggplot(aes(Sample, sample.weights, fill = Genotype)) + 
	geom_bar(stat = "identity") + 
	facet_wrap(~Genotype, scales= "free_x") + 
	geom_hline(yintercept = 1, linetype = 2) +
	scale_y_continuous(expand = expand_scale(c(0, 0.05))) +
	labs(y = "Sample Weight") +
	theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5))
```


### Genotype checks

```{r psen2IDs}
psen2IDs <- subset(transcripts, symbol == "psen2") %>% names()
```

```{r, fig.height=4, fig.cap = "*CPM values for each psen2 transcript across all samples.*"}
transCounts$counts %>%
	cpm() %>% 
	set_rownames(str_remove(rownames(.), "\\.[0-9]+")) %>%
	magrittr::extract(psen2IDs,) %>%
	as.data.frame() %>%
	rownames_to_column("Transcript") %>%
	gather(key = "Sample", value = "CPM", -Transcript) %>%
	mutate(Transcript = str_replace(Transcript, "ENSDART.+", "psen2-WT")) %>%
	left_join(transVoom$targets) %>%
	ggplot(aes(Sample, CPM, fill = Transcript)) +
	geom_bar(stat = "identity") +
	facet_wrap(~Genotype, scales = "free_x") +
	theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5))
```

Transcript abundances (using CPM) were calculated for each of the three *psen2* transcripts, and showed expected patterns of heterozygous expression for FAD samples and all WT expression for the WT samples.
However for sample `8_FS_4`, no WT allele was detected which is quite inexplicable, and this sample should be excluded from all analyses.
The remaining FS samples showed reduced abundance of the FS transcript, as expected under NMD.
No increases in expression of the WT allele were evident, supporting a lack of genetic compensation.

This sample was then removed from all objects.
<!-- , along with `12_WT_4` which had been consistently down-weighted. -->

```{r removeSample}
dropSamples <- c("8_FS_4", "12_WT_4")[1]
transVoom <- transVoom[, !colnames(transVoom) %in% dropSamples]
geneVoom <- geneVoom[, !colnames(geneVoom) %in% dropSamples]
dge <- dge[,!colnames(dge) %in% dropSamples]
```


### Data Inspection

The next step was to perform an MDS analysis.
However, minimal separation was observed between sample groups, 
A simple PCA also revealed that the first few principal components capture less of the total variability than might be expected,

```{r mds}
mds <- geneVoom %>%
	plotMDS(plot = FALSE) %>%
	extract2("cmdscale.out") %>%
	set_colnames(paste0("Dim", 1:2))
```

```{r, echo=FALSE, fig.cap = "*MDS plot showing no clear groups within the data. Point sizes indicate sample weights as calculated by `voomWithQualityWeights()`.*"}
plotly::ggplotly(mds %>%
	cbind(geneVoom$targets[rownames(.), ]) %>%
	ggplot(aes(Dim1, Dim2, colour = Genotype, size = sample.weights, label = Sample)) +
	geom_point() +
	guides(size = FALSE)
)
```

```{r, echo=FALSE}
dge %>%
	cpm(log = TRUE) %>%
	t() %>%
	prcomp() %>%
	summary() %>%
	extract2("importance") %>%
	magrittr::extract(,paste0("PC", 1:5)) %>%
	pander(caption = paste("*First five principal components, showing that the first two only account for", percent(.[3,2]), "of the total variance, which is below expectations*"))
```


## DGE Analysis

### Design

```{r design}
mm <- model.matrix(~0 + Genotype, data = geneVoom$targets) %>%
	set_colnames(gsub("Genotype", "", colnames(.)))
cm <- makeContrasts(FSvWT = FS - WT,
										FADvWT = FAD - WT,
										FADvFS = FAD - FS,
										levels = colnames(mm))
```

Three comparisons were defined with the first two being the difference between the two mutant families and the wild-type samples.
The third comparison was defined as being between the two mutant groups.

```{r geneFit}
geneFit <- geneVoom %>%
	lmFit(mm) %>%
	contrasts.fit(cm) %>%
	eBayes()
```

```{r geneTopTabs}
geneTopTabs <- colnames(cm) %>%
	sapply(function(x){
		topTable(geneFit, x, sort.by = "p", n = Inf) %>%
			rownames_to_column("symbol") %>%
			as_tibble() %>%
			dplyr::rename(FDR = adj.P.Val) %>%
			mutate(Comparison = x,
						 Sig = FDR < 0.05) %>%
			dplyr::select(symbol, logFC, AveExpr, t, P.Value, FDR, Sig)
	}, simplify = FALSE)
```

### FS Vs WT

```{r}
nSig <- geneTopTabs$FSvWT %>%
	dplyr::filter(FDR < 0.05) %>%
	nrow()
```


The first analysis was comparing *psen2^N140fs/+^* samples to *psen2^+/+^* samples.
A total of `r nSig` genes were potentially detected as differentially expressed using an FDR of 5%.
In the following plots, a negative value for logFC corresponds to decreased expression in the heterozygous mutants.

```{r, echo=FALSE, fig.cap="*MD plot for psen2^N140fs/+^ samples compared to psen2^+/+^ samples*"}
geneTopTabs$FSvWT %>%
	arrange(Sig) %>%
	ggplot(aes(AveExpr, logFC, colour = Sig)) +
	geom_point() +
	geom_label_repel(aes(label = symbol), . %>% dplyr::filter(Sig)) +
	scale_colour_manual(values = c(rgb(0.4, 0.4, 0.4, 0.4), "red")) +
	labs(x = "Average Expression (CPM)") +
	ggtitle(expression(paste(italic(psen2^"N140fs/+"), " Vs. ", italic(psen2^"+/+")))) +
	theme(legend.position = "none",
				plot.title = element_text(hjust = 0.5))
```

```{r, echo=FALSE, fig.cap="*Volcano plot for psen2^N140fs/+^ samples compared to psen2^+/+^ samples*"}
geneTopTabs$FSvWT %>%
	arrange(Sig) %>%
	ggplot(aes(logFC, -log10(P.Value), colour = Sig)) +
	geom_point() +
	geom_text_repel(aes(label = symbol), . %>% dplyr::filter(FDR < 0.1 | logFC > 9)) +
	scale_colour_manual(values = c(rgb(0.4, 0.4, 0.4, 0.4), "red")) +
	labs(y = expression(paste(-log[10], "p"))) +
	ggtitle(expression(paste(italic(psen2^"N140fs/+"), " Vs. ", italic(psen2^"+/+")))) +
	theme(legend.position = "none",
				plot.title = element_text(hjust = 0.5))
```

```{r, echo=FALSE}
geneTopTabs$FSvWT %>%
	dplyr::slice(1:10) %>%
	dplyr::select(Symbol = symbol, 
								logFC, AveExpr, P.Value, FDR) %>%
	pander(split.tables = Inf,
				 style = "rmarkdown",
				 justify = "lrrrr",
				 caption = "*10 most highly ranked genes in the comparison between psen2^N140fs/+^ samples and psen2^+/+^ samples*")
```

```{r, echo=FALSE, fig.cap = "*Expression patterns for significantly DE genes in the comparison between psen2^N140fs/+^ samples and psen2^+/+^ samples. Values are given as CPM using an offset of 1 to avoid zeroes, with the y-axis being displayed on the log scale.*"}
cpm(dge) %>% 
	add(1) %>%
	magrittr::extract(geneTopTabs$FSvWT$symbol[1:nSig],) %>%
	as.data.frame() %>%
	rownames_to_column("ID") %>%
	gather(key = "Sample", value = "CPM", -ID) %>%
	as_tibble() %>%
	left_join(dge$samples) %>%
	ggplot(aes(Genotype, CPM, fill = Genotype)) +
	geom_boxplot() +
	scale_y_log10() +
	facet_wrap(~ID, scales = "free_y") +
	labs(y = "CPM + 1")
```

```{r, echo=FALSE, fig.cap = "*Expression patterns for the next most highly ranked genes in the comparison between psen2^N140fs/+^ samples and psen2^+/+^ samples, but which are not formally considered as DE. Values are given as CPM using an offset of 1 to avoid zeroes, with the y-axis being displayed on the log scale.*"}
cpm(dge) %>% 
	add(1) %>%
	magrittr::extract(geneTopTabs$FSvWT$symbol[seq(nSig + 1, length.out = 6)],) %>%
	as.data.frame() %>%
	rownames_to_column("ID") %>%
	gather(key = "Sample", value = "CPM", -ID) %>%
	as_tibble() %>%
	left_join(dge$samples) %>%
	ggplot(aes(Genotype, CPM, fill = Genotype)) +
	geom_boxplot() + 
	scale_y_log10() +
	facet_wrap(~ID, scales = "free_y") +
		labs(y = "CPM + 1")
```

### FAD Vs WT

```{r}
nSig <- geneTopTabs$FADvWT %>%
	dplyr::filter(FDR < 0.05) %>%
	nrow()
```


The next analysis was comparing *psen2^T141_L142delinsMISLISV/+^* samples to *psen2^+/+^* samples.
No genes could be considered as DE using an FDR anywhere up to 50%.
In the following plots, a negative value for logFC corresponds to decreased expression in the heterozygous mutants.

```{r, echo=FALSE, fig.cap="*MD plot for psen2^T141_L142delinsMISLISV/+^ samples compared to psen2^+/+^ samples*"}
geneTopTabs$FADvWT %>%
	arrange(Sig) %>%
	ggplot(aes(AveExpr, logFC, colour = Sig)) +
	geom_point() +
	geom_text_repel(aes(label = symbol),
									. %>% dplyr::filter(logFC > 5)) +
	geom_text_repel(aes(label = symbol),
									. %>% dplyr::filter(logFC < -3.5)) +
	scale_colour_manual(values = c(rgb(0.4, 0.4, 0.4, 0.4), "red")) +
	labs(x = "Average Expression (CPM)") +
	ggtitle(expression(paste(italic(psen2^"T141_L142delinsMISLISV/+"), " Vs. ", italic(psen2^"+/+")))) +
	theme(legend.position = "none",
				plot.title = element_text(hjust = 0.5))
```

```{r, echo=FALSE, fig.cap="*Volcano plot for psen2^T141_L142delinsMISLISV/+^ samples compared to psen2^+/+^ samples*"}
geneTopTabs$FADvWT %>%
	arrange(Sig) %>%
	ggplot(aes(logFC, -log10(P.Value), colour = Sig)) +
	geom_point() +
	geom_text_repel(aes(label = symbol),
									. %>% dplyr::filter(logFC > 5)) +
	geom_text_repel(aes(label = symbol),
									. %>% dplyr::filter(logFC < -3.5)) +
	scale_colour_manual(values = c(rgb(0.4, 0.4, 0.4, 0.4), "red")) +
	scale_x_continuous(breaks = seq(-8, 8, by = 2)) +
	labs(y = expression(paste(-log[10], "p"))) +
	ggtitle(expression(paste(italic(psen2^"T141_L142delinsMISLISV/+"), " Vs. ", italic(psen2^"+/+")))) +
	theme(legend.position = "none",
				plot.title = element_text(hjust = 0.5))
```

```{r, echo=FALSE}
geneTopTabs$FADvWT %>%
	dplyr::slice(1:10) %>%
	dplyr::select(Symbol = symbol, 
								logFC, AveExpr, P.Value, FDR) %>%
		pander(split.tables = Inf,
				 style = "rmarkdown",
				 justify = "lrrrr",
				 caption = "*10 most highly ranked genes in the comparison between psen2^T141_L142delinsMISLISV/+^ samples and psen2^+/+^ samples*")
```

```{r, echo=FALSE, fig.cap = "*Expression patterns for the 5 most highly ranked genes in the comparison between psen2^T141_L142delinsMISLISV/+^ samples and psen2^+/+^ samples. None were considered as DE. Values are given as CPM using an offset of 1 to avoid zeroes, with the y-axis being displayed on the log scale.*"}
cpm(dge) %>% 
	add(1) %>%
	magrittr::extract(geneTopTabs$FADvWT$symbol[1:5],) %>%
	as.data.frame() %>%
	rownames_to_column("ID") %>%
	gather(key = "Sample", value = "CPM", -ID) %>%
	as_tibble() %>%
	left_join(dge$samples) %>%
	ggplot(aes(Genotype, CPM, fill = Genotype)) +
	geom_boxplot() + 
	facet_wrap(~ID, scales = "free_y") +
	scale_y_log10() +
	theme(legend.position = c(5/6, 1/4))
```



### FAD Vs FS 

```{r}
nSig <- geneTopTabs$FADvFS %>%
	dplyr::filter(FDR < 0.05) %>%
	nrow()
```


The final analysis was comparing *psen2^T141_L142delinsMISLISV/+^* samples to *psen2^N140fs/+^* samples.
A total of `r nSig` genes were potentially detected as differentially expressed using an FDR of 5%.
In the following plots, a negative value for logFC corresponds to decreased expression in *psen2^T141_L142delinsMISLISV/+^* samples, whilst a positive value for logFC corresponds to increased expression in *psen2^T141_L142delinsMISLISV/+^* samples.

```{r, echo=FALSE, fig.cap="*MD plot for psen2^T141_L142delinsMISLISV/+^ samples compared to psen2^N140fs/+^ samples*"}
geneTopTabs$FADvFS %>%
	arrange(Sig) %>%
	ggplot(aes(AveExpr, logFC, colour = Sig)) +
	geom_point() +
	geom_label_repel(aes(label = symbol), . %>% dplyr::filter(Sig)) +
	scale_colour_manual(values = c(rgb(0.4, 0.4, 0.4, 0.4), "red")) +
	labs(x = "Average Expression (CPM)") +
	ggtitle(expression(paste(italic(psen2^"T141_L142delinsMISLISV/+"), " Vs. ", italic(psen2^"N140fs/+")))) +
	theme(legend.position = "none",
				plot.title = element_text(hjust = 0.5))
```

```{r, echo=FALSE, fig.cap="*Volcano plot for psen2^T141_L142delinsMISLISV/+^ samples compared to psen2^N140fs/+^ samples*"}
geneTopTabs$FADvFS %>%
	arrange(Sig) %>%
	ggplot(aes(logFC, -log10(P.Value), colour = Sig)) +
	geom_point() +
	geom_text_repel(aes(label = symbol), . %>% dplyr::filter(FDR < 0.1 | abs(logFC) > 6)) +
	scale_colour_manual(values = c(rgb(0.4, 0.4, 0.4, 0.4), "red")) +
	labs(y = expression(paste(-log[10], "p"))) +
	ggtitle(expression(paste(italic(psen2^"T141_L142delinsMISLISV/+"), " Vs. ", italic(psen2^"N140fs/+")))) +
	theme(legend.position = "none",
				plot.title = element_text(hjust = 0.5))
```

```{r, echo=FALSE}
geneTopTabs$FADvFS %>%
	dplyr::slice(1:10) %>%
	dplyr::select(Symbol = symbol, 
								logFC, AveExpr, P.Value, FDR) %>%
	pander(split.tables = Inf,
				 style = "rmarkdown",
				 justify = "lrrrr",
				 caption = "*10 most highly ranked genes in the comparison between psen2^T141_L142delinsMISLISV/+^ samples and psen2^N140fs/+^ samples*")
```

```{r, echo=FALSE, fig.cap = "*Expression patterns for significantly DE genes in the comparison between psen2^T141_L142delinsMISLISV/+^ samples and psen2^N140fs/+^ samples. This is essentially a subset of the previously identified genes*", fig.height=4}
cpm(dge) %>% 
	add(1) %>%
	magrittr::extract(geneTopTabs$FADvFS$symbol[1:nSig],) %>%
	as.data.frame() %>%
	rownames_to_column("ID") %>%
	gather(key = "Sample", value = "CPM", -ID) %>%
	as_tibble() %>%
	left_join(dge$samples) %>%
	ggplot(aes(Genotype, CPM, fill = Genotype)) +
	geom_boxplot() + 
	facet_wrap(~ID, scales = "free_y") +
	scale_y_log10()
```

```{r, echo=FALSE, fig.cap = "*Expression patterns for the next most highly ranked genes in the comparison between psen2^T141_L142delinsMISLISV/+^ samples and psen2^N140fs/+^ samples, but which are not formally considered as DE*"}
cpm(dge) %>% 
	add(1) %>%
	magrittr::extract(geneTopTabs$FADvFS$symbol[seq(nSig + 1, length.out = 5)],) %>%
	as.data.frame() %>%
	rownames_to_column("ID") %>%
	gather(key = "Sample", value = "CPM", -ID) %>%
	as_tibble() %>%
	left_join(dge$samples) %>%
	ggplot(aes(Genotype, CPM, fill = Genotype)) +
	geom_boxplot() + 
	facet_wrap(~ID, scales = "free_y") +
	scale_y_log10() +
	theme(legend.position = c(5/6, 1/4))
```

### Similarity between Mutants

Although there were minimal DE genes in the above analysis, the similarity between differential expression values (i.e. logFC) was inspected visually.

```{r plotMutantComparison, echo=FALSE, fig.cap = "*Comparison between mutants showing logFC for each mutant, based on comparison against WT samples. Genes considered statistically significant between mutants are highlighted in red, however several other genes demonstrated either highly similar behaviour, or were suggestive of different behviours. These genes are labelled in grey. Dahed horizontal and vertical lines have been placed at $\\pm1$, with the unit line also shown in pale blue.*"}
geneFit$coefficients %>%
	as.data.frame() %>%
	rownames_to_column("gene_name") %>%
	mutate(Sig = gene_name %in% dplyr::filter(geneTopTabs$FADvFS, Sig)$symbol) %>%
	ggplot(aes(FSvWT, FADvWT, colour = Sig)) +
	geom_abline(slope = 1, intercept = 0, linetype = 2, colour = rgb(0.8, 0.8, 1)) +
	geom_vline(xintercept = c(-1, 1), linetype = 2, colour = "grey60") +
	geom_hline(yintercept = c(-1, 1), linetype = 2, colour = "grey60") +
	geom_point(alpha = 0.5) +
	geom_text_repel(aes(label = gene_name),
									. %>% dplyr::filter(Sig),
									colour = "red") +
	geom_text_repel(aes(label = gene_name),
									. %>% dplyr::filter(!Sig, FADvWT > 4),
									colour = "grey60") +
	geom_text_repel(aes(label = gene_name),
									. %>% dplyr::filter(!Sig, FADvWT < -3),
									colour = "grey60") +
	geom_text_repel(aes(label = gene_name),
									. %>% dplyr::filter(!Sig, 
																			abs(FSvWT) > 4.8,
																			abs(FADvWT) < 2),
									colour = "grey60") +
	scale_colour_manual(values = c("grey50", "red")) +
	scale_x_continuous(breaks = seq(-8, 8, by = 4), minor_breaks = seq(-10, 10, by = 2)) +
	scale_y_continuous(breaks = seq(-4, 8, by = 4), minor_breaks = seq(-4, 8, by = 2)) +
	labs(x = "logFC (FS Vs WT)",
			 y = "logFC (FAD Vs WT)") +
	theme(legend.position = "none")
```

```{r, echo=FALSE, fig.cap = "*Expression patterns for genes showing similiarity of apparent differential expression across both mutants when compared to WT samples. In all cases one or more outlier points appears to have impacted the ability for these genes to be considered as DE. With the exception of the first two genes, these outliers samples were not consistent. Jitter has been added to the x-axis.*"}
genes2plot <- c("AL929206.1", "BX004838.1", "si:ch211-233m11.1", "si:ch211-181d7.1", "zgc:55461", "BX323045.1")
plotly::ggplotly(
	cpm(dge) %>% 
		add(1) %>%
		magrittr::extract(genes2plot,) %>%
		as.data.frame() %>%
		rownames_to_column("ID") %>%
		gather(key = "Sample", value = "CPM", -ID) %>%
		as_tibble() %>%
		left_join(geneVoom$targets) %>%
		ggplot(aes(Genotype, CPM, colour = Genotype, size = sample.weights, label = Sample)) +
		geom_jitter(width = 0.25) + 
		labs(y = "CPM + 1") +
		facet_wrap(~ID, scales = "free_y") +
		scale_y_log10() +
		theme(legend.position = "none")
)
```

```{r, echo=FALSE, fig.cap = "*Expression patterns for genes showing potential differences in apparent differential expression across both mutants when compared to WT samples. In all cases one or more outlier points appears to have impacted the ability for these genes to be considered as DE. With the exception of the second and third genes in the top row, these outliers samples were not consistent. Jitter has been added to the x-axis.*", fig.height=9}
genes2plot <- c("arrdc1b", "BX324227.3", "BX324227.2", "CR626939.1", "CR381638.1", "si:ch211-76m11.3", "mybpc2b", "BX470069.1")
plotly::ggplotly(
	cpm(dge) %>% 
		add(1) %>%
		magrittr::extract(genes2plot,) %>%
		as.data.frame() %>%
		rownames_to_column("ID") %>%
		gather(key = "Sample", value = "CPM", -ID) %>%
		as_tibble() %>%
		left_join(geneVoom$targets) %>%
		ggplot(aes(Genotype, CPM, colour = Genotype, size = sample.weights, label = Sample)) +
		geom_jitter(width = 0.25) + 
		labs(y = "CPM + 1") +
		facet_wrap(~ID, scales = "free_y") +
		scale_y_log10() +
		theme(legend.position = "none")
)
```


```{r, results='hide'}
names(geneTopTabs) %>%
	lapply(function(x){
		nm <- paste0(x, ".tsv.gz") %>% gzfile("w")
		geneTopTabs[[x]] %>%
			dplyr::select(-Sig) %>%
			write_tsv(nm)
		close(nm)
	})
```

## Enrichment Analysis 

### Hallmark Gene Sets

```{r ens2Entrez}
ens2Entrez <- read_tsv("~/Documents/Zebrafish_RNAseq_DE/data/z11Ensembl_hg38Entrez.tsv") %>%
	set_names(c("GENEID", "EntrezID"))
```

```{r ranks}
ranks <- list(
	FSvWT = genes %>%
		as.data.frame() %>%
		dplyr::select(gene_id, symbol) %>%
		right_join(geneTopTabs$FSvWT) %>%
		as_tibble() %>%
		arrange(t) %>%
		with(structure(t, names = gene_id)),
	FADvWT = genes %>%
		as.data.frame() %>%
		dplyr::select(gene_id, symbol) %>%
		right_join(geneTopTabs$FADvWT) %>%
		as_tibble() %>%
		arrange(t) %>%
		with(structure(t, names = gene_id))
)
```


```{r hallmarkEns}
hallmarkEns <- file.path("~/Documents/Zebrafish_RNAseq_DE/data/", "h.all.v6.2.entrez.gmt") %>%
	gmtPathways() %>%
	mclapply(function(x){
		dplyr::filter(ens2Entrez, EntrezID %in% x)$GENEID %>%
			unique()
	}, mc.cores = 12)
```


```{r hallmarkRes}
set.seed(123)
hallmarkRes <- names(ranks) %>%
	lapply(function(x){
		fgsea(hallmarkEns, ranks[[x]], nperm = 1e5) %>% 
			as_tibble() %>%
			arrange(pval) %>%
			mutate(
				padj = p.adjust(pval, "bonferroni"),
				fdr = p.adjust(pval, "fdr")
			)
	}) %>%
	set_names(names(ranks))
```

```{r}
names(hallmarkRes) %>% 
	lapply(function(x){ mutate(hallmarkRes[[x]], comparison = x)}) %>% 
	bind_rows() %>% 
	dplyr::select(pathway, NES, comparison) %>%
	spread(key = comparison, NES) %>% 
	mutate(pathway = str_remove_all(pathway, "HALLMARK_")) %>% 
	ggplot(aes(FADvWT, FSvWT)) + 
	geom_text_repel(aes(label = pathway)) + 
	geom_point() + 
	geom_abline(slope = 1, intercept = 0, linetype = 2, colour = "grey50") + 
	geom_hline(yintercept = hallmarkRes$FSvWT %>%
						 	dplyr::filter(fdr < 0.05) %>%
						 	.[["NES"]] %>%
						 	abs %>%
						 	min() %>%
						 	multiply_by(c(-1, 1)), 
						 colour = "red", linetype = 2) + 
	geom_vline(xintercept = hallmarkRes$FADvWT %>%
						 	dplyr::filter(fdr < 0.05) %>%
						 	.[["NES"]] %>%
						 	abs %>%
						 	min() %>%
						 	multiply_by(c(-1, 1)), 
						 colour = "red", linetype = 2) +
	annotate("rect", 
					 xmin = hallmarkRes$FADvWT %>%
						 	dplyr::filter(fdr < 0.05) %>%
						 	.[["NES"]] %>%
						 	abs %>%
						 	min() %>%
						 	multiply_by(-1),
					 xmax = hallmarkRes$FADvWT %>%
						 	dplyr::filter(fdr < 0.05) %>%
						 	.[["NES"]] %>%
						 	abs %>%
						 	min(),
					 ymin = hallmarkRes$FSvWT %>%
						 	dplyr::filter(fdr < 0.05) %>%
						 	.[["NES"]] %>%
						 	abs %>%
						 	min() %>%
						 	multiply_by(-1),
					 ymax = hallmarkRes$FSvWT %>%
						 	dplyr::filter(fdr < 0.05) %>%
						 	.[["NES"]] %>%
						 	abs %>%
						 	min(),
					 fill = "grey50",
					 alpha = 0.2
	) +
	labs(x = "Normalised Enrichment Score (FADvWT)",
			 y = "Normalised Enrichment Score (FSvWT)")
```

```{r}
hallmarkRes %>% write_rds("psen2_FSHetVsWt_FADHetVsWt_hallmark_gsea.rds")
```

### KEGG Gene Sets

```{r keggEns}
keggEns <- file.path("~/Documents/Zebrafish_RNAseq_DE/data/c2.cp.kegg.v6.2.entrez.gmt") %>%
	gmtPathways() %>%
	mclapply(function(x){
		dplyr::filter(ens2Entrez, EntrezID %in% x)$GENEID %>%
			unique()
	}, mc.cores = 12)
```


```{r keggRes}
set.seed(123)
keggRes <- names(ranks) %>%
	lapply(function(x){
		fgsea(keggEns, ranks[[x]], nperm = 1e5) %>% 
			as_tibble() %>%
			arrange(pval) %>%
			mutate(
				padj = p.adjust(pval, "bonferroni"),
				fdr = p.adjust(pval, "fdr")
			)
	}) %>%
	set_names(names(ranks))
```

```{r, eval = FALSE}
names(keggRes) %>% 
	lapply(function(x){ mutate(keggRes[[x]], comparison = x)}) %>% 
	bind_rows() %>% 
	dplyr::select(pathway, NES, comparison) %>%
	spread(key = comparison, NES) %>% 
	mutate(pathway = str_remove_all(pathway, "KEGG_")) %>% 
	ggplot(aes(FADvWT, FSvWT)) + 
	geom_text_repel(aes(label = pathway)) + 
	geom_point() + 
	geom_abline(slope = 1, intercept = 0, linetype = 2, colour = "grey50") + 
	geom_hline(yintercept = keggRes$FSvWT %>%
						 	dplyr::filter(fdr < 0.05) %>%
						 	.[["NES"]] %>%
						 	abs %>%
						 	min() %>%
						 	multiply_by(c(-1, 1)), 
						 colour = "red", linetype = 2) + 
	geom_vline(xintercept = keggRes$FADvWT %>%
						 	dplyr::filter(fdr < 0.05) %>%
						 	.[["NES"]] %>%
						 	abs %>%
						 	min() %>%
						 	multiply_by(c(-1, 1)), 
						 colour = "red", linetype = 2) +
	annotate("rect", 
					 xmin = keggRes$FADvWT %>%
						 	dplyr::filter(fdr < 0.05) %>%
						 	.[["NES"]] %>%
						 	abs %>%
						 	min() %>%
						 	multiply_by(-1),
					 xmax = keggRes$FADvWT %>%
						 	dplyr::filter(fdr < 0.05) %>%
						 	.[["NES"]] %>%
						 	abs %>%
						 	min(),
					 ymin = keggRes$FSvWT %>%
						 	dplyr::filter(fdr < 0.05) %>%
						 	.[["NES"]] %>%
						 	abs %>%
						 	min() %>%
						 	multiply_by(-1),
					 ymax = keggRes$FSvWT %>%
						 	dplyr::filter(fdr < 0.05) %>%
						 	.[["NES"]] %>%
						 	abs %>%
						 	min(),
					 fill = "grey50",
					 alpha = 0.2
	) +
	labs(
		x = "Normalised Enrichment Score (FADvWT)",
		y = "Normalised Enrichment Score (FSvWT)"
	)
```

```{r}
keggRes %>% write_rds("psen2_FSHetVsWt_FADHetVsWt_kegg_gsea.rds")
```



## Differential Transcript Expression

As the level of transcript complexity is less in zebrafish than human, and 1:1 mapping between species is less robust, only a brief analysis was performed at the transcript level.
In essence, the same genes were found as the most highly ranked, with changes in expression of *psen2* transcripts detected as expected, providing a form of positive control.
Following the top tables, the basic transcript expression patterns are shown for three possible genes of interest.
Notably, the transcripts showing the strongest differential expression are expressed at very low-levels for both *si:ch211-132g1.3* and *slc37a4b*.

```{r transFit}
transFit <- transVoom %>%
	lmFit(mm) %>%
	contrasts.fit(cm) %>%
	eBayes()
```

```{r transTopTabes}
transTopTabs <- colnames(cm) %>%
	sapply(function(x){
		topTable(transFit, x, sort.by = "p", n = Inf) %>%
			as_tibble() %>%
			dplyr::rename(FDR = adj.P.Val) %>%
			mutate(Comparison = x,
						 Sig = FDR < 0.05,
						 Region = paste0(as.character(seqnames), ":", start, "-", end, ":", as.character(strand))) %>%
		dplyr::select(tx_id, symbol, logFC, AveExpr, t, P.Value, FDR, gene_id, Region, tx_biotype, Sig)
	}, simplify = FALSE)
```

```{r echo = FALSE}
transTopTabs$FSvWT %>%
	dplyr::slice(1:10) %>%
	dplyr::select(Transcript = tx_id, Symbol = symbol,
								logFC, AveExpr, P.Value, FDR, gene_id) %>%
	pander(split.tables = Inf,
				 style = "rmarkdown",
				 justify = "llrrrrl",
				 caption = "*10 most highly ranked transcripts in the comparison between psen2^N140fs/+^ samples and psen2^+/+^ samples*")
```

```{r, echo = FALSE}
transTopTabs$FADvWT %>%
	dplyr::slice(1:10) %>%
	dplyr::select(Transcript = tx_id, Symbol = symbol,
								logFC, AveExpr, P.Value, FDR, gene_id) %>%
	pander(split.tables = Inf,
				 style = "rmarkdown",
				 justify = "llrrrrl",
				 caption = "*10 most highly ranked transcripts in the comparison between psen2^T141_L142delinsMISLISV/+^ samples and psen2^+/+^ samples*")
```


```{r, echo = FALSE}
transTopTabs$FADvFS %>%
	dplyr::slice(1:10) %>%
	dplyr::select(Transcript = tx_id, Symbol = symbol,
								logFC, AveExpr, P.Value, FDR, gene_id) %>%
	pander(split.tables = Inf,
				 style = "rmarkdown",
				 justify = "llrrrrl",
				 caption = "*10 most highly ranked transcripts in the comparison between psen2^T141_L142delinsMISLISV/+^ samples and psen2^N140fs/+^ samples*")
```


```{r ENSDARG00000089477, echo = FALSE, fig.width=10, fig.cap = "One transcript (ENSDART00000137332) was undetectable in the FS mutants, and this is a non-coding transcript."}
transCounts$counts %>% 
	cpm() %>%
	add(1) %>%
	magrittr::extract(
		dplyr::filter(transVoom$genes, gene_id == "ENSDARG00000089477")$tx_id_version,) %>%
	as.data.frame() %>% 
	rownames_to_column("tx_id") %>% 
	gather(key = "Sample", value = "CPM", -tx_id) %>% 
	left_join(transVoom$targets) %>% 
	dplyr::filter(!is.na(sample.weights)) %>% 
	mutate(tx_id = str_remove(tx_id, "\\.[0-9]+")) %>%
	ggplot(aes(tx_id, CPM, fill = Genotype)) + 
	geom_boxplot() +
	scale_y_log10() +
	labs(x = "Transcript ID",
			 y = "CPM + 1") +
	ggtitle("Transcripts for ENSDARG00000089477 (si:ch211-132g1.3)") +
	theme(plot.title = element_text(hjust = 0.5))
```

```{r}
subset(transcripts, gene_id == "ENSDARG00000089477") %>%
	subset(tx_name %in% rownames(transVoom)) %>%
	.[order(.$tx_biotype, decreasing = TRUE), c("tx_id", "tx_biotype")] %>% 
	as.data.frame() %>% 
	pander(split.tables = Inf, 
				 captions = "Detected transcripts for ENSDARG00000089477 (si:ch211-132g1.3)")
```


```{r ENSDARG00000015540, echo = FALSE}
transCounts$counts %>% 
	cpm() %>%
	add(1) %>%
	magrittr::extract(
		dplyr::filter(transVoom$genes, gene_id == "ENSDARG00000015540")$tx_id_version,) %>%
	as.data.frame() %>% 
	rownames_to_column("tx_id") %>% 
	gather(key = "Sample", value = "CPM", -tx_id) %>% 
	left_join(transVoom$targets) %>% 
	dplyr::filter(!is.na(sample.weights)) %>% 
	mutate(tx_id = str_remove(tx_id, "\\.[0-9]+")) %>%
	ggplot(aes(tx_id, CPM, fill = Genotype)) + 
	geom_boxplot() +
	scale_y_log10() +
	labs(x = "Transcript ID",
			 y = "logCPM") +
	ggtitle("Transcripts for ENSDARG00000015540 (psen2)") +
	theme(plot.title = element_text(hjust = 0.5))
```

```{r ENSDARG00000077180, echo = FALSE, fig.width=10}
transCounts$counts %>% 
	cpm() %>%
	add(1) %>%
	magrittr::extract(
		dplyr::filter(transVoom$genes, gene_id == "ENSDARG00000077180")$tx_id_version,) %>%
	as.data.frame() %>% 
	rownames_to_column("tx_id") %>% 
	gather(key = "Sample", value = "CPM", -tx_id) %>% 
	left_join(transVoom$targets) %>% 
	dplyr::filter(!is.na(sample.weights)) %>% 
	mutate(tx_id = str_remove(tx_id, "\\.[0-9]+")) %>%
	ggplot(aes(tx_id, CPM, fill = Genotype)) + 
	geom_boxplot() +
	scale_y_log10() +
	labs(x = "Transcript ID",
			 y = "logCPM") +
	ggtitle("Transcripts for ENSDARG00000077180 (slc37a4b)") +
	theme(plot.title = element_text(hjust = 0.5))
```

## Session Info

```{r sessionInfo, echo=FALSE}
pander(sessionInfo())
```

